blog/content/post/2021-12-02-meta-nezha.md

```diff
@@ -0,0 +1,162 @@
+the recipe's code isn't that clear and for now, it isn't prepared to be
+upstreamed.
+
+## Nezha D1 meta layer construction
+
+Nezha D1 layer is using the following Yocto Project meta layers in `hardknott`
+version:
+
+* [poky](https://git.yoctoproject.org/git/poky)
+
+* [meta-openembedded](https://git.openembedded.org/meta-openembedded)
+
+* [meta-riscv](https://github.com/riscv/meta-riscv.git)
+
+The structure of a `meta-nezha` is divided into two main parts:
+
+* `meta-nezha-bsp` - this layer contains recipes for `boot0`, `u-boot`,
+  `OpenSBI` and `linux kernel`. It also contains machine configuration which
+  set/unset or enable/disable the key features of the board,
+
+* `meta-nezha-distro` - contain a recipe for the minimal image, configuration of the
+  system eg. what init manager is used and `wks` file which is used to create
+  image file which can be flashed to the SD card.
+
+Key repositories used by `meta-nezha-bsp` recipes are forks of repositories
+patched / created by **smaeul**:
+
+* [OpenSBI](https://github.com/Cezarus27/opensbi/tree/d1-wip): `d1-wip` branch
+
+* [u-boot](https://github.com/Cezarus27/u-boot/tree/d1-wip): `d1-wip` branch
+
+* [Linux](https://github.com/Cezarus27/linux/tree/riscv/d1-wip): `riscv/d1-wip`
+  branch
+
+* [boot0](https://github.com/Cezarus27/sun20i_d1_spl): `mainline` branch
+
+> Note: I decide to fork the `smaeul` repositories because the original repos
+> are often updated and rebased to the newest branches from upstream
+> repositories and it often caused problems with the proper build of the yocto
+> image.
+
+Main work to do was to adopt the bootflow of Nezha to the Yocto Project. In this
+process, the `boot0` recipe was created from the ground. That wasn't so smooth
+to integrate `boot0` compilation flow with the Yocto build engine due to missing
+some headers. So the patch for `boot0` Makefile was provided also and it is
+applied during the building process.
+
+It wasn't enough because a lot of work had to be done to compile `u-boot` for
+the D1 chip. `u-boot-nezha` recipe contains procedures of creating the TOC1
+image from binaries described in `toc.cfg`. Creating the TOC image is different
+than others because this image contains three binaries which one is from
+`OpenSBI`. So it was necessary to add a dependency for `u-boot` on `OpenSBI`
+during the compilation task. The `tftp-mmc-boot.txt` script was also modified to
+provide `bootargs` specific for the Nezha.
+
+The situation with Linux kernel recipe isn't different. Due to the "work in
+progress" status of the support for D1 chip in Linux kernel, I decided to create
+a separate recipe of the kernel for the Nezha board. It provides to the building
+system two configuration files with kernel options that enable `autofs4` and `cgroups`. Enabling these options was necessary because, without them, kernel
```

Comment:
wrapping, also, the same issue as for U-Boot... maybe we could contribute some fixes?

Reply:
I think so

Reply:
@Cezarus27 did you try to create some issue or PR for that?

Reply:
Not yet. I think only the patch for U-Boot, for now, is valuable to be upstreamed. Kernel configuration (`autofs4` and `cgroups`) I saw in Fedora configuration:
* https://github.com/tekkamanninja/linux/tree/allwinner_nezha_d1_devel/arch/riscv/configs